### DISCUSSION TITLE
store master database on disk instead of RAM

### DISCUSSION THREAD

**Discussion by ethaniel** [moosefs/moosefs#394] [Ideas]
I am new to MooseFS and considering deploying it, however I am saddened by it's requirement of keeping the whole master database in RAM (which would amount to 50+ gb in my case). I've been looking into alternatives (like BeeGFS) which seem to be storing the master metadata on disks instead. Would such a thing be available to moosefs anytime soon?

**Comment by jkiebzak** [comment]
#231 asks the same question. I believe there is another gitlab issue asking
something similar.

Currently no way to do this, since the focus is on performance.

>

**Comment by pkonopelko** [comment]
Hello @ethaniel,
As @jkiebzak wrote and as it was mentioned in #231, such approach would cause a dramatic performance degradation. You can see it yourself if your Master Server start swapping, even if the swap space was stored on a NVMe disk. Master Server keeps all the metadata's data structures in RAM to minimize the latency necessary to query it for metadata. Also RAM is quite cheap these days.

How did you count that your Master Server would require 50+ GB of RAM? How many files are you going to store on your cluster? Master Server utilizes about 50 GiB of RAM if you have about 230 million of files on your cluster. Of course it may vary a bit depending on the replication goal, whether you need (or not) extra metadata like `xattrs`, etc. But anyway number of files/directories/..., or rather: objects in the files system is a key parameter here.

PS: Alternatives like mentioned BeeGFS doesn't provide a built-in redundancy like replication goal in MooseFS (they rely on underlying redundant storage), and more.

Piotr

**Comment by smurfix** [comment]
Well, performance would degrade if you do it badly. But let's face it, most archival file systems have a few hot spots that get accessed daily, and a really large long tail which gets accessed … well, basically never. Thus I seriously doubt that on-disk metadata storage plus a large(-enough) LRU cache would hurt anybody.

In fact, *every* file system does this, it's just that the caching is done by the kernel.

**Comment by chogata** [comment]
Well, in MooseFS there is no metadata that "never" gets accessed. It's a distributed, fault tolerant system. All components constantly self test and any errors found are repaired as soon as possible - lost chunks are replicated, etc. In a big system, there is always something going on, even with data the user thinks is not used at all. Law of big numbers: if a disk fails after 5 years of usage, when you have 300 of them, one will fail every week ;) I/O errors happen often. And anything that is not accessible, not readable, is soon repaired. This would not be possible with enough efficiency if metadata was mostly stored on HDDs.

Also, archive isn't the only application of MooseFS. Many people use it to keep a lot of data that is constantly accessed - even, if it is not written/modified often, it is read often. Performance is crucial here and simply cannot be achieved to such degree without keeping metadata in RAM.

MooseFS is designed to be fast and safe - our focus is on that.
